### PR TITLE
feat: map upgrades

### PR DESCRIPTION
**Pull request by MaxwellDG**
Apologies for the huge PR, but I found myself rewriting the same pages for a variety of different github issues. Dealing with the repeated merge conflicts would've been equally as frustrating.

This addresses:  #2975 #2955 #2894 and several other features discussed in person that add-on to #2830 

**Initial loading**
- If location permissions given -> initial loading is at user's location
- else if it's user's first time using map -> initial loading at centre of user's country's ip
- else -> initial loading at last viewed coordinates from previous session

**Speed**
- Map loading speed decreased to ~2 seconds from ~8 seconds on Android

**Asking permissions**
- Added `NSLocationAlwaysAndWhenInUseUsageDescription` permission to Info.plist so Apple will shutup. Even though we're not asking for that permission as shown by line 47 in maps-screen.tsx
`Geolocation.setRNConfiguration({
  skipPermissionRequests: true,
  enableBackgroundLocationUpdates: false,
  authorizationLevel: "whenInUse",
  locationProvider: "auto",
})`

- Permissions are no longer asked immediately upon visiting the map
- Added a button in top-right corner that looks like 'center-on user location' button. When pressed, it will ask for user's permission to access location
- If user has previously pressed 'Never ask again', then we can't ask for permissions. We can however open a modal that brings them directly to their system's Blink settings screen where they can update the permission themselves

**iOS bugfix**
- tapping a marker that's close to another marker would make it jump to the wrong one. now if that happens, it'll stay on the correct marker. It'll flash though. Definitely a big improvement, yet not perfect.

**TODO**
- ~~Only part not tested on iOS is the new button for asking permissions~~
- ~~latLng in cache should be an object not 2 separate values. syntax is confusing here~~
- ~~On some Android devices, the 2nd time permissions are asked for (only the 2nd time ever), the modal asking them to change it in their settings will popup after they say no. Will make separate issue for this cause I'm stumped at the moment.~~

### CODE REVIEW

File: ios/Podfile [hunk 1]
```diff
@@ -1,14 +1,26 @@
...
+node_require('react-native-permissions/scripts/setup.rb')
 
 use_modular_headers!
 
 platform :ios, min_ios_version_supported
 prepare_react_native_project!
+
+setup_permissions([
+  'LocationAccuracy',
+  'LocationAlways',
```
Comment (nicolasburtey) [inline ios/Podfile:20]: we only need permission when device is in used, could this be the permission for running in the background?

File: app/components/map-component/index.tsx [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+  handleCalloutPress,
+  onError,
+}: Props) {
+  const {
+    theme: { colors, mode: themeMode },
+  } = useTheme()
+  const styles = useStyles()
+  const { LL } = useI18nContext()
+  const client = useApolloClient()
+  const text = React.useMemo(() => LL.MapScreen.payBusiness(), [LL.MapScreen])
```
Comment (nicolasburtey) [inline app/components/map-component/index.tsx:49]: why is that necessary? it's not a pattern we have used so far to cache translated text

File: app/components/map-component/index.tsx [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+      }
+      // Android can ask twice for permission, and initial checks cannot differentiate between BLOCKED vs DENIED
+    } else {
+      // TODO figure out how to not fire this after the 2nd time user blocks
+      toggleModal()
+    }
+  }
+
+  const centerOnUser = () => {
+    getUserRegion(async (region) => {
```
Comment (nicolasburtey) [inline app/components/map-component/index.tsx:74]: is this an async action?

File: app/components/map-component/index.tsx [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+  data,
+  userLocation,
+  permissionsStatus,
+  setPermissionsStatus,
+  handleMapPress,
+  handleMarkerPress,
+  focusedMarker,
+  focusedMarkerRef,
+  handleCalloutPress,
+  onError,
```
Comment (nicolasburtey) [inline app/components/map-component/index.tsx:41]: does `onError` needs to be pass as a prop? is it only ever specific to `alertOnLocationError`?

File: app/components/map-component/index.tsx [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+            e.nativeEvent.coordinate.longitude ===
+            focusedMarker?.mapInfo.coordinates.longitude
+          if (!matchingLat || !matchingLng) {
+            if (focusedMarkerRef.current) {
+              focusedMarkerRef.current.showCallout()
+            }
+          }
+        }}
+      >
+        {(data?.businessMapMarkers ?? []).reduce(
```
Comment (nicolasburtey) [inline app/components/map-component/index.tsx:137]: this PR will help using a .map instead: https://github.com/GaloyMoney/galoy/pull/3856

**Comment by nicolasburtey** [comment]
on android simulator, this is the first thing I see when I click on the map

image

is it intended?

**Comment by MaxwellDG** [comment]
> on android simulator, this is the first thing I see when I click on the map
> 
> image
> is it intended?

No, that should only be shown when pressing the button in the top right corner. I haven't tested on android simulator. I'll check that out now.

**Comment by nicolasburtey** [comment]
nevermind, it seems my app was not correctly updated.

File: app/components/map-component/index.tsx [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    request(LOCATION_PERMISSION)
+      .then((status) => {
+        if (status === RESULTS.GRANTED) {
+          centerOnUser()
+        } else if (status === RESULTS.BLOCKED) {
+          respondToBlocked(status)
+        }
+        setPermissionsStatus(status)
+      })
+      .catch(() => alertOnLocationError())
```
Comment (nicolasburtey) [inline app/components/map-component/index.tsx:93]: could we conver this to async/await syntax?
we don't really then().catch(), imo it's less readable than the new code syntax

File: app/graphql/client-only-query.ts [hunk 7]
```diff
@@ -41,6 +44,13 @@ export default gql`
     countryCode @client
   }
 
+  query latLng {
+    latLng @client {
+      lat
+      lng
```
Comment (nicolasburtey) [inline app/graphql/client-only-query.ts:50]: I think we should also store the zoom. because otherwise we come back in a very "unzoomed" settings by default, and it feels almost like the settings was not saved as a result of it

File: app/screens/map-screen/map-screen.tsx [hunk 8]
```diff
@@ -95,168 +120,120 @@ export const MapScreen: React.FC<Props> = ({ navigation }) => {
...
-          if (data) {
-            const region: Region = {
-              latitude: data.coords.latitude,
-              longitude: data.coords.longitude,
-              latitudeDelta: 0.02,
-              longitudeDelta: 0.02,
-            }
-            callback(region)
+        setInitialLocation(region)
+        // User is using maps for the first time, so we center on the center of their IP's country
```
Comment (nicolasburtey) [inline app/screens/map-screen/map-screen.tsx:172]: is this comment for the else branch below? if that is the case, I would move this line below the else, to be consistent with [comment line 163](https://github.com/GaloyMoney/galoy-mobile/pull/2999/files#diff-2840b3b05c86d0fba4f9261d1ba0aaeca6798d1ad37be559b44495d847710f04R163) where the comment is within the block, not before it

File: app/screens/map-screen/map-screen.tsx [hunk 9]
```diff
@@ -95,168 +120,120 @@ export const MapScreen: React.FC<Props> = ({ navigation }) => {
     toastShow({ message: error.message, LL })
   }
 
-  const maps = data?.businessMapMarkers ?? []
+  // On screen load, check (NOT request) if location permissions are given
+  React.useEffect(() => {
+    check(LOCATION_PERMISSION).then((status) => {
```
Comment (nicolasburtey) [inline app/screens/map-screen/map-screen.tsx:125]: async/await syntax preferred

File: app/screens/map-screen/map-screen.tsx [hunk 10]
```diff
@@ -95,168 +120,120 @@ export const MapScreen: React.FC<Props> = ({ navigation }) => {
...
+      console.log("Initial check: ", status)
```
Comment (nicolasburtey) [inline app/screens/map-screen/map-screen.tsx:126]:
```suggestion
```

**Comment by nicolasburtey** [comment]
generally the PR looks great. biggest comment is on the zoom not been stored across session. otherwise it's mostly some nits. I have not review every line throughly but I think that is fine - this is not a critical code like the payment flow that require deep attention.

**Comment by MaxwellDG** [comment]
@nicolasburtey Added zoom, addressed the small refactors.

I also found a solution for that Android 2nd request edge case. Found the location where I could add a flag.

**Review by nicolasburtey — APPROVED** [review]
